subiquity/models/filesystem.py

```diff
@@ -115,8 +116,91 @@ def asdict(inst):
 # in the FilesystemModel or FilesystemController classes.
 
 
+class DeviceAction(enum.Enum):
+    INFO = enum.auto()
+    EDIT = enum.auto()
+    PARTITION = enum.auto()
+    FORMAT = enum.auto()
+    DELETE = enum.auto()
+
+
+@attr.s
+class _Formattable:
+    # Base class for anything that can be formatted and mounted,
+    # e.g. a disk or a RAID or a partition.
+
+    # Filesystem
+    _fs = attr.ib(default=None, repr=False)
+    # Nothing yet, but one day RAID, LV, ZPool, BCache...
+    _constructed_device = attr.ib(default=None, repr=False)
+
+    def _is_entirely_used(self):
+        return self._fs is not None or self._constructed_device is not None
+
+    def fs(self):
+        return self._fs
+
+    def constructed_device(self):
+        return self._constructed_device
+
+
 @attr.s
-class Disk:
+class _Device(_Formattable):
+    # Anything that can have partitions, e.g. a disk or a RAID.
+
+    # subclass must implement .size!
```

Comment:
Should this be an Abstract  Base class then?

Reply:
Yeah, I'll change it. Was confused about how to do abstract properties but on looking it up it's really easy.